utilities/env_librados.cc

```diff
@@ -0,0 +1,961 @@
+        */
+    {
+        ret = _rados.ioctx_create(_pool_name.c_str(), _io_ctx);
+        if (ret < 0) {
+            std::cerr << "couldn't set up ioctx! error " << ret << std::endl;
+            ret = EXIT_FAILURE;
+            goto out;
+        }
+    }
+
+    /* add root dir */
+    _AddFid(ROOT_DIR_TAG, DIR_TAG_VALUE);
+
+    out:
+    LOG_DEBUG("rados connect result code : %i\n", ret);
+}
+
+
+/* Get fid store in meta object (the object that store all metadata information). DIR also store in meta object, but fid of DIR is <DIR> */
+std::string EnvLibrados::_CreateFid() {
+    return _db_id + "." + GenerateUniqueId();
+}
+
+rocksdb::Status EnvLibrados::_GetFidByOmap(
+        const std::string &fname, 
+        std::string& fid) {
+    LOG_DEBUG(" %s\n", fname.c_str());
+    std::set<std::string> keys;
+    std::map<std::string, librados::bufferlist> kvs;
+    keys.insert(fname);
+    int r = _io_ctx.omap_get_vals_by_keys(_db_id, keys, &kvs);
+    LOG_DEBUG("%i|%lld\n", r, (long long)kvs.size());
+
+    if (0 == r && 0 == kvs.size()) {
+        return rocksdb::Status::NotFound();
+    } else if (0 == r && 0 != kvs.size()) {
+        fid = kvs[fname].to_str();
+        LOG_DEBUG(" %s\n", fid.c_str());
+        return rocksdb::Status::OK();
+    } else {
+        return err_to_status(r);
+    }
+}
+rocksdb::Status EnvLibrados::_AddFidByOmap(
+        const std::string& fname, 
+        const std::string& fid) {
+    std::map<std::string, librados::bufferlist> kvs;
+    librados::bufferlist value;
+    value.append(fid);
+    kvs[fname] = value;
+    int r = _io_ctx.omap_set(_db_id, kvs);
+    LOG_DEBUG("%s|%s|%i\n", fname.c_str(), fid.c_str(), r);
+    return err_to_status(r);
+}
+
+/*
+    RocksDB has a 2-level structure, so all keys that have dir as prefix are subfiles of dir
+ */
+rocksdb::Status EnvLibrados::_GetSubFnamesByOmap(
+    const std::string& dir,
+    std::vector<std::string> * result
+    ) {
+    std::string start_after(dir);
+    std::string filter_prefix(dir);
+    std::map<std::string, librados::bufferlist> kvs;
+    _io_ctx.omap_get_vals(_db_id, 
+        start_after, filter_prefix, 
+        MAX_ITEMS_IN_FS, &kvs);
+
+    result->clear();
+    for (auto i = kvs.begin(); i != kvs.end(); i++) {
+        LOG_DEBUG("[match]%s\n", i->first.c_str());
+        result->push_back(i->first.substr(dir.size()+1));
+    }
+    return rocksdb::Status::OK();
+}
+
+rocksdb::Status EnvLibrados::_DelFidByOmap(
+    const std::string& fname) {
+    LOG_DEBUG("%s\n", fname.c_str());
+    std::set<std::string> keys;
+    keys.insert(fname);
+    int r = _io_ctx.omap_rm_keys(_db_id, keys);
+    return err_to_status(r);
+}
+
+
+
+rocksdb::Status EnvLibrados::_GetFid(
+        const std::string &fname, 
+        std::string& fid) {
+    return _GetFidByOmap(fname, fid);
+}
+rocksdb::Status EnvLibrados::_AddFid(
+        const std::string& fname, 
+        const std::string& fid) {
+    return _AddFidByOmap(fname, fid);
+}
```

Comment:
I think these can be collapsed... _GetFid etc and drop the *ByOmap variations/subfunctions